pkg/gameserversets/gameserversets.go

```diff
@@ -15,22 +15,83 @@
 package gameserversets
 
 import (
-	stablev1alpha1 "agones.dev/agones/pkg/apis/stable/v1alpha1"
+	"sort"
+
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
 	listerv1alpha1 "agones.dev/agones/pkg/client/listers/stable/v1alpha1"
 	"github.com/pkg/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 )
 
+// node is just a convenience data structure for
+// keeping relevant GameServer information about Nodes
+type node struct {
+	name  string
+	total int64
+	ready []*v1alpha1.GameServer
+}
+
+// filterGameServersOnLeastFullNodes returns a limited list of GameServers, ordered by the nodes
+// they are hosted on, with the least utilised Nodes being prioritised
+func filterGameServersOnLeastFullNodes(gsList []*v1alpha1.GameServer, limit int32) []*v1alpha1.GameServer {
```

Comment:
@victor-prodan I took my best shot at optimising this, but you always have a good eye for algorithm optimisations - any thoughts?

Reply:
It had been a very busy week for me, I will have a look this weekend

Reply:
No dramas. End of year I'm not travelling much, so I have more time to write code and not be on airplanes. Take your time. :+1: